src/biokbase/narrative/jobs/job.py

```diff
@@ -77,65 +117,104 @@ def __init__(self, **kwargs):
+        )
 
-        self.update_state(ee2_state)
+        return cls(
+            ee2_state=ee2_state,
+            extra_data=kwargs.get("extra_data"),
+            children=kwargs.get("children")
+        )
 
-    @classmethod
-    def from_state(cls, job_state):
+    def __getattr__(self, name):
         """
-        Parameters:
-        -----------
-        job_state - dict
-            the job information returned from ee2.check_job
-
+        Map expected job attributes to paths in stored ee2 state
         """
-        job_input = job_state.get("job_input", {})
-        narr_cell_info = job_input.get("narrative_cell_info", {})
-        return cls(
-            app_id=job_input.get("app_id", JOB_DEFAULTS["app_id"]),
-            app_version=job_input.get("service_ver", JOB_DEFAULTS["app_version"]),
-            batch_id=job_state.get("batch_id", JOB_DEFAULTS["batch_id"]),
-            batch_job=job_state.get("batch_job", JOB_DEFAULTS["batch_job"]),
-            cell_id=narr_cell_info.get("cell_id", JOB_DEFAULTS["cell_id"]),
-            child_jobs=job_state.get("child_jobs", JOB_DEFAULTS["child_jobs"]),
-            ee2_state=job_state,
-            job_id=job_state.get("job_id"),
-            owner=job_state.get("user", JOB_DEFAULTS["owner"]),
-            params=job_input.get("params", JOB_DEFAULTS["params"]),
-            run_id=narr_cell_info.get("run_id", JOB_DEFAULTS["run_id"]),
-            tag=narr_cell_info.get("tag", JOB_DEFAULTS["tag"]),
+        attr = dict(
+            app_id=lambda: self._const_state.get("job_input", {}).get("app_id", JOB_ATTR_DEFAULTS["app_id"]),
+            app_version=lambda: self._const_state.get("job_input", {}).get("service_ver", JOB_ATTR_DEFAULTS["app_version"]),
+            batch_id=lambda: (
+                self.job_id if self.batch_job else
+                self._const_state.get("batch_id", JOB_ATTR_DEFAULTS["batch_id"])
+            ),
+            batch_job=lambda: self._const_state.get("batch_job", JOB_ATTR_DEFAULTS["batch_job"]),
+            cell_id=lambda: self._const_state.get("job_input", {}).get("narrative_cell_info", {}).get("cell_id", JOB_ATTR_DEFAULTS["cell_id"]),
+            child_jobs=lambda: self.state().get("child_jobs", JOB_ATTR_DEFAULTS["child_jobs"]),
+            job_id=lambda: self._const_state.get("job_id"),
+            params=lambda: self._const_state.get("job_input", {}).get("params", JOB_ATTR_DEFAULTS["params"]),
+            retry_ids=lambda: self.state().get("retry_ids", JOB_ATTR_DEFAULTS["retry_ids"]),
+            retry_parent=lambda: self._const_state.get("retry_parent", JOB_ATTR_DEFAULTS["retry_parent"]),
+            run_id=lambda: self._const_state.get("job_input", {}).get("narrative_cell_info", {}).get("run_id", JOB_ATTR_DEFAULTS["run_id"]),
+            tag=lambda: self._const_state.get("job_input", {}).get("narrative_cell_info", {}).get("tag", JOB_ATTR_DEFAULTS["tag"]),
+            user=lambda: self._const_state.get("user", JOB_ATTR_DEFAULTS["user"]),
```

Comment:
I think I see what's happening. So if you were to run, for example
```
job.app_id
```
it should look first in `job._const_state["job_input"]["app_id"]` and return the default otherwise. So, avoiding having to do a bunch of default calculation elsewhere.

I think that's probably ok.